gms/gossiper.cc

```diff
@@ -2150,27 +2275,38 @@ future<> gossiper::wait_for_gossip(std::chrono::milliseconds initial_delay, std:
 }
 
 future<> gossiper::wait_for_gossip_to_settle() {
-    static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
-    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
+    logger.info("Waiting for gossip to settle with query...");
     auto do_enable_features = [this] {
         return async([this] {
             if (!std::exchange(_gossip_settled, true)) {
                maybe_enable_features();
             }
         });
     };
+    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
     if (force_after == 0) {
         return do_enable_features();
     }
-    return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS, force_after).then([this, do_enable_features] {
```

Comment:
If we boot a cluster with some nodes down, how long is `wait_for_cluster_ring_status_to_match()` expected to wait before failing?

It will eventually throw, and we'll proceed to the old way which waits GOSSIP_SETTLE_MIN_WAIT_MS. So we will have a regression in restart time.

The purpose of the wait here is not to sync

Reply:
Let me add to that:

we should only `wait_for_cluster_ring_status_to_match` when **joining** a node to the cluster. And in case of failure, we should not fall-back to the old method: we should abort the joining procedure. Otherwise, the node we can't contact during joining won't learn about our tokens, so it will send writes to the wrong replicas.

In the other case, when we already are in the cluster and we're just restarting, we should NOT do this, because it should always be safe to restart a node irrespective of whether we can contact others or not (of course today it's not true because we don't assume that a node didn't join when we restart; it could have joined without us knowing while we were down; problem is, the joining node couldn't have known whether we were down or just behind a network partition)

Reply:
> If we boot a cluster with some nodes down, how long is `wait_for_cluster_ring_status_to_match()` expected to wait before failing?
> 
> It will eventually throw, and we'll proceed to the old way which waits GOSSIP_SETTLE_MIN_WAIT_MS. So we will have a regression in restart time.
> 
> The purpose of the wait here is not to sync

```
 525             if (nodes_down) {
 526                 logger.info("Wait for cluster_ring_status, matched={}, needed={}, nodes_down={}, nodes_to_check={}, gave up due to nodes down",
 527                         matched, needed, nodes_down, nodes_to_check);
 528                 throw std::runtime_error("Gave up wait for gossip with query becasue nodes are down");
 529             }
```

If we found a node down, we gave up immediately and fallback. We do not wait for a timeout at all. So, no regression.

Reply:
Kamil, when existing boots up, we still want to do the wait_for_cluster_ring_status_to_match check. We want the booting node to see the same cluster view as the result of the cluster, and we want to avoid the timeout which slows down the boot process.

Reply:
But what if we want to join a node and it can't contact some other node in the cluster? Should we allow it to join? (the other node might be up and running)

Reply:
> 
> If we found a node down, we gave up immediately and fallback. We do not wait for a timeout at all. So, no regression.

How long does it take for the RPC call to figure out that the node is down?

Reply:
> Let me add to that:
> 
> we should only `wait_for_cluster_ring_status_to_match` when **joining** a node to the cluster. And in case of failure, we should not fall-back to the old method: we should abort the joining procedure. Otherwise, the node we can't contact during joining won't learn about our tokens, so it will send writes to the wrong replicas. That's the "making boot process safer" part of this PR, if I understand correctly.
> 
> In the other case, when we already are in the cluster and we're just restarting, we should NOT do this, because it should always be safe to restart a node irrespective of whether we can contact others or not (of course today it's not true because we don't assume that a node didn't join when we restart; it could have joined without us knowing while we were down; problem is, the joining node couldn't have known whether we were down or just behind a network partition, so it shouldn't have joined at all)

I agree that it's safest if joining fails with anode down, and then restarts don't have to do anything. 

We're not there yet because there is a controversy about failing join with nodes down. Until that's resolved, we need to attempt to sync on restart to reduce risk.

Reply:
@kbr- I agree with you on this.

Reply:
@tgrabiec great, maybe we will be able to convince others and reach consensus at some point. I also have this ace in my sleeve which I'm keeping in a frame on my desk (the dicussion was made in the context of a 1000-node cluster):
![image](https://user-images.githubusercontent.com/6516501/77331429-fed56380-6d20-11ea-94bd-8b020e2be2ed.png)
(I hope @avikivity didn't change his mind about this)
There's still the last part about "relaxing the restriction", however, I think we should first implement a restricted, but *provably safe* solution, and then consider how to relax the restriction without losing safety. I've been doing some research on this lately and unfortunately I don't think it can be done in a completely asynchronous model of distributed computation, but maybe it can be with some weak partial synchrony assumptions (which can be made "in practice") with some kind of [leases](https://en.wikipedia.org/wiki/Lease_(computer_science)). We could also depend on the administrator to "promise that a node is really down", and let joining proceed in that case, but administrators have a tendency to shot themselves in their feet. In any case, relaxing should be a discussion for the future; today I'm fighting for safety.

Reply:
> > If we found a node down, we gave up immediately and fallback. We do not wait for a timeout at all. So, no regression.
> 
> How long does it take for the RPC call to figure out that the node is down?

If a node is down (the port 7000 is not listened) it is immediate. 

In addition to that, the rpc verb is set to timeout in 5 seconds.

Reply:
With repair based node operations, we do not have to sync data from the node that is losing the range. We can sync data from other replicas, e.g., quorum of replicas. This means a down node does not prevent the joining node.

Again, how to add node or remove into / from  the cluster, when to allow, when to reject areoutside the scope of this series.

Reply:
> > > If we found a node down, we gave up immediately and fallback. We do not wait for a timeout at all. So, no regression.
> > 
> > 
> > How long does it take for the RPC call to figure out that the node is down?
> 
> If a node is down (the port 7000 is not listened) it is immediate.
> 
> In addition to that, the rpc verb is set to timeout in 5 seconds.

So if the node is unreachable, this patch will make the boot slower by 5 seconds, doubling the wait inside `wait_for_gossip_to_settle()`. Maybe not a big problem, but avoidable if we use a deadline in the fallback method calculated on entry as `now + 5s`.

Reply:
The 5s timeout happens only when the node is slow to respond. I think in this case, it is totally fine with the 5s wait and then fallback to old method. 

I am not sure what you mean by ` but avoidable if we use a deadline in the fallback method calculated on entry as now + 5s`

Do you mean the new method fallback to old method if the query exceeds 5s dead line?
This is what we do now in this patch already.

Reply:
wait_for_gossip() accepts an `initial_delay`. I suggested to make it a time point instead of duration so that the time spent on the new wait is subtracted from that initial delay:

```
start = now()
return new_wait_for_gossip().handle_exception([start, &] {
    return wait_for_gossip(start + GOSSIP_SETTLE_MIN_WAIT_MS, force_after)'
});
```

But it's not that important.

Reply:
@asias 
> we do not have to sync data from the node that is losing the range. We can sync data from other replicas, e.g., quorum of replicas. This means a down node does not prevent the joining node.

That's not true. The node that is losing the range might be the only node that contains a write.
Example 1: someone might have performed a write with CL = ONE which landed on this node only. Then a CL = ALL read should see the write. If you don't stream from this node, the write will be lost, and CL = ALL reads won't see it after bootstrapping finishes.
Example 2: someone might have performed a write with a higher CL (say QUORUM), but the write could have "failed" and reached only this node. Some later reads observed this write (for example, CL = ALL reads). Again we shouldn't loose this write due to a node bootstrapping, we should provide read monotonicity.

It doesn't matter if you use streaming or repair, that's an implementation detail. What matters is which replicas you get the data from and which replicas you send the data to.

Reply:
In example1, since you write with CL = one,  if you lose a node, you lose the replica. That's it.  In this case, we can not bootstrap a new node. 

In example2, the write should fail if the write reach only one node. 

It matters because with stream we only be able to stream from one of the node, with repair we can be more flexible. 

If we always read and write with quorum, we can bootstrap with one node down.

Reply:
> In example1, since you write with CL = one, if you lose a node, you lose the replica. That's it.

But I did not lose a node. I bootstrapped another node, the old node stopped being a replica, the new node started being a replica. The old node was previously the only replica which had this data. The data should have been transferred to the new node.

> In example2, the write should fail if the write reach only one node.

It did fail, but some later reads observed the data (because the write did reach the node).

>  If we always read and write with quorum

But you can't assume that. We offer other consistency levels.

Reply:
> > In example1, since you write with CL = one, if you lose a node, you lose the replica. That's it.
> 
> But I did not lose a node. I bootstrapped another node, the old node stopped being a replica, the new node started being a replica. The old node was previously the only replica which had this data. The data should have been transferred to the new node.

Yes, I understand the issue. In this case, we can not bootstrap a node.

> > In example2, the write should fail if the write reach only one node.
> 
> It did fail, but some later reads observed the data (because the write did reach the node).

This happens with/without bootstrapping a node.  What can we do?

> > If we always read and write with quorum
> 
> But you can't assume that. We offer other consistency levels.

Sure. I am talking about the common cases.

Reply:
> Yes, I understand the issue. In this case, we can not bootstrap a node.

Why? We can --- we simply need to make sure to stream data from the replica that will stop being a replica.
If you're saying that we cannot bootstrap a node if this replica is down --- that is correct.
My point is, you never know if this replica contains some data by itself. So you should always make sure it is up before you bootstrap.

> This happens with/without bootstrapping a node. What can we do?

We still provide certain consistency guarantees, even if the write failed.
For example, with quorum reads, we provide read monotonicity: if you performed a read and this read observed some data, later reads should also observe this data, even if this data came from a failed write. This is achieved with read repair if you use QUORUM, and it just works if you use ALL. We should make sure we don't break this property while bootstrapping.

> Sure. I am talking about the common cases.

The problem is: we don't know during bootstrapping whether all writes were made using quorum or not. And even if they were, we still need to ensure that ALL reads observe ONE writes.